File: src/go/problems/count_characters_test.go
// https://leetcode.com/problems/find-words-that-can-be-formed-by-characters

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type CountCharacters struct{}

func TestCountCharacters(t *testing.T) {
	gen := core.TestSuite[CountCharacters]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]string{"cat","bt","hat","tree"}).Param("atach").Result(6)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]string{"hello","world","leetcode"}).Param("welldonehoneyr").Result(10)
	}).Run(t)
}

func (CountCharacters) Solution(words []string, chars string) int {
    charMap := make([]int, 26)

	for _, ch := range chars {
		charMap[ch - 'a']++
	}

	result := 0
	for _, word := range words {
		wordMap := make([]int, 26)
		matched := true

		for _, ch := range word {
			wordMap[ch - 'a']++

			if wordMap[ch - 'a'] > charMap[ch - 'a'] {
				matched = false
				break
			}
		}

		if matched {
			result += len(word)
		}
	}

	return result
}

File: src/python/problems/count_characters_test.py
# https://leetcode.com/problems/find-words-that-can-be-formed-by-characters

from core.problem_base import *

class CountCharacters(ProblemBase):
    def Solution(self, words: list[str], chars: str) -> int:
        charMap = {}

        for ch in chars:
            charMap[ch] = charMap.get(ch, 0) + 1

        result = 0
        for word in words:
            wordMap = {}
            matched = True

            for ch in word:
                wordMap[ch] = wordMap.get(ch, 0) + 1

                if ch not in charMap or wordMap[ch] > charMap[ch]:
                    matched = False
                    break

            if matched:
                result += len(word)

        return result

if __name__ == '__main__':
    TestGen(CountCharacters) \
        .Add(lambda tc: tc.Param(["cat","bt","hat","tree"]).Param("atach").Result(6)) \
        .Add(lambda tc: tc.Param(["hello","world","leetcode"]).Param("welldonehoneyr").Result(10)) \
        .Run()
